rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,227 @@
+
+# RFC 005: Soft Chain Upgrades
+
+## Changelog
+
+- 2020-11-16: Initial draft.
+
+## Author(s)
+
+- Callum Waters (@cmwaters)
+
+## Context
+
+The current method for upgrading a network across block protocol breaking
+changes is to hard fork the chain. All nodes must coordinate a halt height/
+time, where application state can be exported across and an entirely new chain
+(as far as Tendermint is concerned) is created (see [cosmoshub-3
+upgrade](https://blog.cosmos.network/cosmos-hub-3-upgrade-announcement-39c9da941aee)).
+This means that each block protocol version maps to the entire life of a single
+chain. The result of this is that either Tendermint is hindered with the
+flexibility and quantity of changes it can make or that networks have to deal
+with pain of frequently upgrading.
+
+Furthermore, although work from [RFC002: Non-Zero Genesis](https://github.com/tendermint/spec/blob/master/rfc/002-nonzero-genesis.md)
+has enabled unique heights throughout the history of a "logical" chain, data
+retrievability across heights remains difficult as blockchains that choose to
+upgrade not only have to coordinate upgrades but they essentially lose all their
+history.
+
+The scope of this RFC is to address the need for Tendermint to have flexibility
+in improving itself whilst offering greater ease of use to the networks running
+on Tendermint. This is done by introducing a division in the types of ways a
+network can be upgraded. There will still remain the need for hard upgrades (the
+starting of a new chain) but this will not be focused on in this
+RFC(discussions on improving this can be found [here](https://github.com/tendermint/tendermint/issues/5595)).
+Instead, this RFC will only explore what supporting soft upgrades would entail.
+
+## Proposal
+
+Tendermint supports a class of upgrades deemed as soft upgrades.
+
+Soft and hard upgrades would directly align with semver. For example, a release
+from 1.2 → 1.3 would be considered a soft upgrade and could be executed on a
+live chain (more on how that would be executed further down). A major release
+from 1.5 → 2.0 would be a hard upgrade and require a network to halt and migrate.
+
+### Protocol Versioning
+
+Tendermint currently has three protocol versions: Block, P2P and App.
+
+The block version consists of all block related data structures i.e. Header,
+Vote, Validator, Commit. This would be the sole focus for soft upgrades (a full
+list of the exact data structures can be found in
+[data structures](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md#data-structures).
+A block version could be incremented between minor releases. However, for
+backwards compatibility, all prior versions in a given major release would also
+need to be supported thus TM 1.3 could support block version 9 and 10 and TM 1.4
+could introduce block version 11 but would still need to support 9 and 10. In a
```

Comment:
Block version changes are associated with heights. Is it still breaking if a client request a block at a height 20 with version 1 and the block at that height is actually version 2 so the rpc just returns a version error. This basically means that clients still work but they have an invisible ceiling of heights they can query until they upgrade.

> Make sure the data structures of all new block versions can be represented as all prior supported versions, and that clients will have to opt-in to the new data structures by bumping e.g. the RPC version number of requests.

How feasible do you think this is? Isn't this much the same as just making it backwards-compatible like you suggested by having sermver for block versions.

Reply:
> Is it still breaking if a client request a block at a height 20 with version 1 and the block at that height is actually version 2 so the rpc just returns a version error.

Yes, I think so. This will break e.g. block explorers and coin exchanges, and basically all clients that interact with Tendermint that haven't upgraded.

>How feasible do you think this is? Isn't this much the same as just making it backwards-compatible like you suggested by having sermver for block versions.

Depends on what we need. I don't think it buys us much over just "regular" backwards-compatibility, but it might reduce the amount of cruft in newer block versions. E.g. if we rename a field, with backwards compatibility we would have to keep both the old and the new field, while with the other approach we could just map the new field onto the old field for clients that request a previous version and not have the old field present in new versions.

Reply:
Can we enforce that all block version changes are always backwards compatible? i.e. we would never change an existing field from say v10 -> v11, only add fields. I do agree using semver in block versioning as well could be very useful.

Reply:
Yes I think semver versioning makes sense because I think there will also be times where we want to make changes that are not backwards compatible. This is all a juggle between how much freedom we want to give us as developers and how annoying we are to users. 

Quick question: Is adding fields backwards compatible?  If I were to add callums hash to the header in v11 and an external client were to request a block using v10 then this very important hash would be omitted. Is this acceptable?

Reply:
> This is all a juggle between how much freedom we want to give us as developers and how annoying we are to users.

Yup. It's worth bearing in mind that software exists for its users, so they should take priority.

> Is adding fields backwards compatible?

Generally yes, with a couple of important caveats:

* The field must be optional, and the default value must preserve existing behavior.

* The field cannot affect other fields - in this context, notably, it cannot be included in any existing hashes, since clients will no longer be able to verify those hashes (unless it's very clearly specified and implemented up-front how new fields should be hashed, e.g. by using all Protobuf fields in ascending order).

Reply:
But in the context of the block protocol where almost everything is connected and nothing is really optional, I'd imagine it is quite difficult to add a field that is backwards compatible.

Reply:
No-one said this would be easy. :)